### PR TITLE
[shardformer] Pytree fix

### PR DESCRIPTION
**Pull request by CjhHa1**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.
This PR fixes pytree flatten caused bugs

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/pipeline/schedule/_utils.py [hunk 1]
```diff
@@ -118,15 +149,18 @@ def merge_batch(data: List[Any]) -> Any:
...
     tree_spec = None
     for d in data:
-        elems, tree_spec = tree_flatten(d)
+        # elems should be an instance of OrderedDict
+        elems, tree_spec = tree_flatten_hf(d)
         flattened_data.append(elems)
     merged_data = []
-    for elem_batch in zip(*flattened_data):
+
+    for elem_batch in zip(*(flattened_data)):
```
Comment (ver217) [inline colossalai/pipeline/schedule/_utils.py:157]: What's the effect of the brackets?

File: colossalai/pipeline/schedule/one_f_one_b.py [hunk 2]
```diff
@@ -285,5 +296,6 @@ def forward_backward_step(self,
                 self.send_backward(input_obj_grad)
 
         if outputs is not None:
-            outputs = merge_batch(outputs)
+            bsz_dim = 1 if model.unwrap().__class__.__name__ in CHATGLM_MODELS else 0
```
Comment (ver217) [inline colossalai/pipeline/schedule/one_f_one_b.py:299]: Don't judge model name or class. Set this attribute to model and use `getattr()` here.

Comment (CjhHa1) [reply colossalai/pipeline/schedule/one_f_one_b.py]: DONE

Comment (CjhHa1) [reply colossalai/pipeline/schedule/one_f_one_b.py]: move this attribute at `configure`

File: colossalai/pipeline/schedule/_utils.py (hunk 1 above)
Comment (CjhHa1) [reply colossalai/pipeline/schedule/_utils.py]: actually no, it is for other testing and I will remove it

File: colossalai/booster/plugin/hybrid_parallel_plugin.py [hunk 3]
```diff
@@ -27,6 +27,7 @@
 from .pp_plugin_base import PipelinePluginBase
 
 DP_AXIS, PP_AXIS, TP_AXIS = 0, 1, 2
+CHATGLM_MODELS = ['ChatGLMModel', 'ChatGLMForConditionalGeneration']
```
Comment (ver217) [inline colossalai/booster/plugin/hybrid_parallel_plugin.py:30]: Remove this line

Comment (CjhHa1) [reply colossalai/booster/plugin/hybrid_parallel_plugin.py]: DONE

File: colossalai/pipeline/schedule/_utils.py [hunk 4]
```diff
@@ -118,15 +149,18 @@ def merge_batch(data: List[Any]) -> Any:
     flattened_data = []
     tree_spec = None
     for d in data:
-        elems, tree_spec = tree_flatten(d)
+        # elems should be an instance of OrderedDict
+        elems, tree_spec = tree_flatten_hf(d)
```
Comment (ver217) [inline colossalai/pipeline/schedule/_utils.py:153]: Do you need to update `_detach()` with `tree_flatten_hf()`?

Comment (CjhHa1) [reply colossalai/pipeline/schedule/_utils.py:153]: this is done with `tree_map_hf` ,in line 10 ,and the corresponding changes also in one_f one_b

**Review by ver217 — APPROVED** [review]

File: colossalai/pipeline/schedule/_utils.py [hunk 5]
```diff
@@ -118,15 +169,18 @@ def merge_batch(data: List[Any]) -> Any:
     flattened_data = []
     tree_spec = None
     for d in data:
-        elems, tree_spec = tree_flatten(d)
+        print('d: ', d)
```
Comment (ver217) [inline colossalai/pipeline/schedule/_utils.py:172]: Please remove this

Comment (CjhHa1) [reply colossalai/pipeline/schedule/_utils.py:172]: will push a new version after test is done!

Comment (CjhHa1) [reply colossalai/pipeline/schedule/_utils.py]: done!

**Review by ver217 — APPROVED** [review]

**Pull request merged**